### PR TITLE
build: [MLG-305] Check for supported go version at make.

### PR DESCRIPTION
**Pull request by wes-turner**
## Description
Determined cannot be currently built with go1.20. This adds a target in the Makefile that checks the system go version against the supported go version and errors if the system go is not supported.

## Test Plan

This was tested by hand against go1.18.10, go1.19.5, and go1.20.

## Commentary (optional)
This is just a simple fix for an existing problem. It temporizes any decisions about more generally checking requirements, and also leaves for a future commit updates to the library to make determined buildable with more recent go.

## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

MLG-305

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,3 +1,5 @@
+SHELL := bash
```
Comment (dzhu) [inline Makefile:1]: This doesn't seem to be necessary.

File: Makefile [hunk 2]
```diff
@@ -18,10 +20,20 @@ get-deps-pip:
...
 	$(MAKE) get-deps-proto
 
+# Go versions may look like goM, goM.N, or goM.N.P. Only 1.18.* is supported.
+supported_go_minor_version = go1.18
+system_go_version := $(shell go version | sed 's/.*\(go[[:digit:]][[:digit:]0-9.]*\).*/\1/')
+.PHONY: go-version-check
+go-version-check:
+	@: $(if $(findstring $(supported_go_minor_version), $(system_go_version)), \
+				, \
+				$(error go version $(system_go_version) not supported. Must use $(supported_go_minor_version).x))
```
Comment (dzhu) [inline Makefile:35]: This certainly gets the job done, though I must say I was envisioning something a bit more straightforward like `@go version | grep -q 'go1\.18' || (echo 'need go 1.18!' && exit 1)`. Either way works.

File: Makefile [hunk 3]
```diff
@@ -18,10 +20,20 @@ get-deps-pip:
...
 .PHONY: get-deps-go
 get-deps-go:
+	$(MAKE) go-version-check
 	$(MAKE) get-deps-master
 	$(MAKE) get-deps-agent
 	$(MAKE) get-deps-proto
 
+# Go versions may look like goM, goM.N, or goM.N.P. Only 1.18.* is supported.
+supported_go_minor_version = go1.18
+system_go_version := $(shell go version | sed 's/.*\(go[[:digit:]][[:digit:]0-9.]*\).*/\1/')
```
Comment (dzhu) [inline Makefile:30]: Is the `0-9` necessary?

**Review by dzhu — APPROVED** [review]

File: Makefile (hunk 1 above)
Comment (wes-turner) [reply Makefile:1]: That's fair. I'd copied it from `master/Makefile`, where I didn't see any reason to prefer `bash` over `/bin/sh` there, either, just thinking that it's what people expected.

I also like it better without. Removing.

Thanks!

File: Makefile (hunk 3 above)
Comment (wes-turner) [reply Makefile:30]: Thanks for catching that!

File: Makefile (hunk 2 above)
Comment (wes-turner) [reply Makefile:35]: I think I prefer this maybe-overengineering approach to the maybe-underengineering approach. The sed regex is awful, but so long as it's trustable, it just feels more "right" to treat make as a language and perform as much logic in that as possible.

But I'm very much unfamiliar with make idioms. If you're confident that the simpler approach better reflects how people expect make to be used, I'll totally make this change.

Comment (wes-turner) [reply Makefile:35]: Mmm. Speaking of that awful regex, I'm noticing I kind of chose something in-between. Regexes in make aren't great, but I could parse out the go version there from the string returned by `go version`. Worth it?

**Review by dzhu — APPROVED** [review]
I think the style ends up looking perfectly reasonable. Go ahead!

**Pull request merged**